kvs/include/client.hpp

```diff
@@ -0,0 +1,773 @@
+      invalidate_cache_for_worker(worker);
+
+      return try_request(request, trial_limit - 1);
+    }
+
+    KeyTuple tuple = response.tuples(0);
+    if (check_tuple(tuple)) {
+      try_request(request, trial_limit - 1);
+    }
+
+    return response;
+  }
+
+  /**
+   * A helper method to check for the default failure modes for a request that
+   * retrieves a response. It returns true if the caller method should reissue
+   * the request (this happens if errno == 2). Otherwise, it returns false. It
+   * invalidates the local cache if the information is out of date.
+   */
+  bool check_tuple(KeyTuple tuple) {
+    Key key = tuple.key();
+    if (tuple.error() == 2) {
+      logger_->info(
+          "Server ordered invalidation of key address cache for key {}. "
+          "Retrying request.",
+          key);
+
+      invalidate_cache_for_key(key, tuple);
+      return true;
+    }
+
+    if (tuple.has_invalidate() && tuple.invalidate()) {
+      invalidate_cache_for_key(key, tuple);
+
+      logger_->info(
+          "Server ordered invalidation of key address cache for key {}", key);
+    }
+
+    return false;
+  }
+
+  /**
+   * When a server thread tells us to invalidate the cache for a key it's
+   * because we likely have out of date information for that key; it sends us
+   * the updated information for that key, and update our cache with that
+   * information.
+   */
+  void invalidate_cache_for_key(Key key, KeyTuple tuple) {
+    key_address_cache_.erase(key);
+    std::unordered_set<Address> new_cache;
+
+    for (const Address address : tuple.addresses()) {
+      new_cache.insert(address);
+    }
+
+    key_address_cache_[key] = new_cache;
+  }
+
+  /**
+   * Invalidate the key caches for any key that previously had this worker in
+   * its cache. The underlying assumption is that if the worker timed out, it
+   * might have failed, and so we don't want to rely on it being alive for both
+   * the key we were querying and any other key.
+   */
+  void invalidate_cache_for_worker(Address worker) {
+    std::vector<std::string> tokens;
+    split(worker, ':', tokens);
+    std::string signature = tokens[1];
+    std::unordered_set<Key> remove_set;
+
+    for (const auto& key_pair : key_address_cache_) {
+      for (const std::string& address : key_pair.second) {
+        std::vector<std::string> v;
+        split(address, ':', v);
+
+        if (v[1] == signature) {
+          remove_set.insert(key_pair.first);
+        }
+      }
+    }
+
+    for (const std::string& key : remove_set) {
+      key_address_cache_.erase(key);
+    }
+  }
+
+  /**
+   * Prepare a data request object by populating the request ID, the key for
+   * the request, and the response address. This method modifies the passed-in
+   * KeyRequest and also returns a pointer to the KeyTuple contained by this
+   * request.
+   */
+  KeyTuple* prepare_data_request(KeyRequest& request, Key& key) {
```

Comment:
I think it looks cleaner to pass in the request type as an argument to `prepare_data_request` and set the type inside the method.